lib/SPIRV/SPIRVReader.cpp

```diff
@@ -1527,9 +1527,15 @@ Value *SPIRVToLLVM::transValueWithoutDecoration(SPIRVValue *BV, Function *F,
   case OpUndef:
     return mapValue(BV, UndefValue::get(transType(BV->getType())));
 
-  case OpVariable: {
-    auto *BVar = static_cast<SPIRVVariable *>(BV);
-    auto *PreTransTy = BVar->getType()->getPointerElementType();
+  case OpVariable:
+  case OpUntypedVariableKHR: {
+    auto *BVar = static_cast<SPIRVVariableBase *>(BV);
+    SPIRVType *PreTransTy = BVar->getType()->getPointerElementType();
+    if (BVar->getType()->isTypeUntypedPointerKHR()) {
+      auto *UntypedVar = static_cast<SPIRVUntypedVariableKHR *>(BVar);
+      if (SPIRVType *DT = UntypedVar->getDataType())
+        PreTransTy = DT;
```

Comment:
Is it possible for 1536 to fail?  
If it fails then PreTransTy will get the old def `PreTransTy = BVar->getType()->getPointerElementType()` 
but `BVar->getType()->isTypeUntypedPointerKHR()` which seems odd.

Reply:
By specification, we can have untyped variable without data type. Guess I need to add test for global variable without type to see if it's valid

Reply:
I've added test that shows it works as expected. 
That's because `PreTransTy = BVar->getType()->getPointerElementType()` returns the same `OpTypeUntypedPointerKHR` for `OpTypeUntypedPointerKHR` type.

Reply:
From my understanding, when we are handling OpUntypedVariableKHR,
`SPIRVType *PreTransTy = BVar->getType()->getPointerElementType();`
handles the case when the untyped variable has no data type.   I guess this is because getPointerElementType() returns the correct type when there is no data type.
What will PreTransTy be set to if the untyped variable does have a data type?  I am hoping that getPointerElementType() will also return the correct data type, and thus lines 1534 to 1538 can be removed.

Reply:
The idea seems good at first glance, but unfortunately it's not doable and we still do need to have the special condition for untyped variable.

Return type of `OpVariable` is a pointer to the variable type, so it's enough to call `BVar->getType()->getPointerElementType()` to get it.
However, the return type of `OpUntypedVariableKHR` is untyped pointer, which does not hold information about data type by its definition. The only place where we can get info about data type (if it's not empty) is from `UntypedVar->getDataType()`, so the condition can't be removed.